internal/pkg/agent/application/coordinator/coordinator.go

```diff
@@ -547,6 +559,108 @@ func (c *Coordinator) ReExec(callback reexec.ShutdownCallbackFn, argOverrides ..
 	c.reexecMgr.ReExec(callback, argOverrides...)
 }
 
+// Migrate migrates agent to a new cluster and ACKs success to the old one.
+// In case of failure no ack is performed and error is returned.
+func (c *Coordinator) Migrate(ctx context.Context, action *fleetapi.ActionMigrate, backoffFactory func(done <-chan struct{}) backoff.Backoff) error {
+	if !c.isManaged {
+		return ErrNotManaged
+	}
+
+	// check if agent is FS, fail if so
+	if c.isFleetServer() {
+		return ErrFleetServer
+	}
+
+	// Keeping all enrollment options that are not overridden via action
+	options, err := computeEnrollOptions(ctx, paths.ConfigFile(), paths.AgentConfigFile())
+	if err != nil {
+		return fmt.Errorf("failed to compute enroll options: %w", err)
+	}
+
+	persistentConfig, err := enroll.LoadPersistentConfig(paths.ConfigFile())
+	if err != nil {
+		return err
+	}
+
+	originalRemoteConfig, err := options.RemoteConfig(false)
+	if err != nil {
+		return fmt.Errorf("failed to construct original remote config: %w", err)
+	}
+
+	originalClient, err := fleetapiClient.NewAuthWithConfig(
+		c.logger, options.EnrollAPIKey, originalRemoteConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create original fleet client: %w", err)
+	}
```

Comment:
Nit: could we move this block down to just before `originalClient` is used?  Makes it a bit easier to follow the code and also avoids executing this code if something else fails before it.  I realize that means renaming `options` to something like `originalOptions` and `newOptions` but I actually think that's even more readable as it fits the rest of the pattern.